### PR TITLE
[RLlib] Add PPO multi-agent StatelessCartPole learning tests.

### PR DESCRIPTION
**Pull request by sven1977**
Add PPO multi-agent StatelessCartPole learning tests to CI.

* single (CPU) Learner
* 2 CPUs
* single GPU Learner
* 2 GPUs

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/learner/learner_group.py [hunk 1]
```diff
@@ -944,20 +932,23 @@ def __del__(self):
...
-            num_total_mini_batches += 1
-        return num_total_mini_batches
+    def _compute_num_total_mini_batches(
+        episodes,
+        num_shards,
+        mini_batch_size,
+        num_iters,
+    ):
+        # Count total number of timesteps per module ID.
+        if isinstance(episodes[0], MultiAgentEpisode):
```
Comment (simonsays1980) [inline rllib/core/learner/learner_group.py:942]: Why isn't it possible to use generator/iterator that can run empty and if so the learner returns?

File: rllib/env/single_agent_episode.py [hunk 2]
```diff
@@ -1498,54 +1500,92 @@ def slice(self, slice_: slice) -> "SingleAgentEpisode":
...
-            id_=self.id_,
-            # In the following, offset `start`s automatically by lookbacks.
-            observations=InfiniteLookbackBuffer(
-                data=self.get_observations(
-                    slice(start - self.observations.lookback, stop + 1, step),
-                    neg_index_as_lookback=True,
-                ),
-                lookback=self.observations.lookback,
-                space=self.observation_space,
+        _lb = (
```
Comment (simonsays1980) [inline rllib/env/single_agent_episode.py:1503]: Could we leave a comment what we are calculating here and when this case can occur, please?

File: rllib/env/single_agent_episode.py [hunk 3]
```diff
@@ -1498,54 +1500,92 @@ def slice(self, slice_: slice) -> "SingleAgentEpisode":
...
             ),
-            action_space=self.action_space,
-            rewards=InfiniteLookbackBuffer(
-                data=self.get_rewards(
-                    slice(start - self.rewards.lookback, stop, step),
+            lookback=_lb,
+            space=self.action_space,
+        )
+
+        _lb = (
```
Comment (simonsays1980) [inline rllib/env/single_agent_episode.py:1550]: Also could we refactor into a helper function?

File: rllib/tuned_examples/ppo/multi_agent_stateless_cartpole_ppo.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+config = (
+    PPOConfig()
+    # Enable new API stack and use EnvRunner.
+    .api_stack(
+        enable_rl_module_and_learner=True,
+        enable_env_runner_and_connector_v2=True,
+    )
+    .environment("multi_stateless_cart")
+    .env_runners(
+        env_to_module_connector=lambda env: MeanStdFilter(multi_agent=True),
```
Comment (simonsays1980) [inline rllib/tuned_examples/ppo/multi_agent_stateless_cartpole_ppo.py:36]: I remember, we have still this open question, if we need to add this connector also to learner or not. I think we do not need it:  MeanStd rewrites observations and learner has these observations then, correct?

**Review by simonsays1980 — APPROVED** [review]
LGTM. Would be nice to have some comments and a refactoring into helper functions.

File: rllib/tuned_examples/ppo/multi_agent_stateless_cartpole_ppo.py (hunk 4 above)
Comment (sven1977) [reply rllib/tuned_examples/ppo/multi_agent_stateless_cartpole_ppo.py:36]: Correct, this connector (and most other env-to-module ones) directly writes back into the episode, thus making the change to the observation permanent. Hence no need to also add it to the Learner pipeline as the Learner pipeline then operates on the already changed episodes/observations.

**Pull request merged**